### ISSUE TITLE
Ability to query interfaces that a union implements

### ISSUE THREAD

**Issue by madebysid** [remorses/genql#44]
We're using `genql`'s typings on an upcoming app and its been working really well! It fits right in, and typed queries and mutations are awesome!

There's one area where type support is limited though, and that has to do with interfaces and unions.
When querying against a resource that returns a union, you should also be able to query interfaces that the union's types implement. Currently, it is only possible to query individual types within the union.

Perhaps an example will make this clearer.

So assume this (part of) GraphQL schema

```
interface ClientError {
  message: String!
}

type ClientErrorNameAlreadyTaken implements ClientError {
  message: String!
}

type ClientErrorNameInvalid implements ClientError {
  message: String!
}

union GenericError =
    ClientErrorNameAlreadyTaken
  | ClientErrorNameInvalid

type Query {
  error: GenericError
}
```

Here, this query is currently possible:
```
query q {
  error {
    ... on ClientErrorNameAlreadyTaken { message }
    ... on ClientErrorNameInvalid { message }
  }
}
```
as
```
client.query({
  error: {
    on_ClientErrorNameAlreadyTaken: { message: true }
    on_ClientErrorNameInvalid: { message: true }
  }
})
```

However, the following is also a valid query in GraphQL that `genql` disallows:
```
query q {
  error {
    ... on ClientError { message }
  }
}
```

It'd be awesome if `genql` could let me query it this way:
```
client.query({
  error: {
    on_ClientError: { message: true }
  }
})
```

If I'm being unclear, please let me know!

I'm more than happy to send a PR your way to support this if you like :)

**Comment by remorses** [comment]
This is not that simple to implement, why don't you use the interface directly instead of the union?

```gql
interface ClientError {
  message: String!
}

type ClientErrorNameAlreadyTaken implements ClientError {
  message: String!
}

type ClientErrorNameInvalid implements ClientError {
  message: String!
}

type Query {
  error: ClientError
}
```

**Comment by madebysid** [comment]
So in our case, we'd like to eventually have some error-specific fields as well (the example was rather simplified). So let's say `ClientErrorNameAlreadyTaken` could return a suggested name as well.

The goal of my second query then, is to fetch _common_ fields from a all types within the union.

I'll also let @jasonkuhrt chime in if he wants

**Comment by jasonkuhrt** [comment]
@remorses Because not every operation will return every kind of client error, but we want all client errors to share a base interface.

**Comment by jasonkuhrt** [comment]
> This is not that simple to implement

Would you be open for a PR from us?

**Comment by remorses** [comment]
I will work on this, I already know what to change, i also need to clean up tests because now it's a mess

**Comment by remorses** [comment]
Released in 2.4.0

**Comment by madebysid** [comment]
@remorses Just tried this out, works exactly as expected! Thanks so much for getting this out so incredibly quickly!
